design-document.md

```diff
@@ -232,6 +232,352 @@ Note: Two non-colluding aggregators --- one leader and one standard aggregator
+batch-process a number of inputs at once.
+
+Each helper share has a corresponding message payload contained in
+`PAVerifyStartReq.shares`. This includes any protocol-specific information the
+helper needs for input validation, e.g., the joint randomness used by the leader
+and helper for the protocol run.
+
+Note that a well-formed message contains as many payloads as shares, i.e., the
+sequence `PAVerifyStartReq.payloads` has the same number of elements as
+`PAVerifyStartReq.shares`. Moreover, the i-th payload should correspond to the
+i-th share for each i.
+
+**Response.**
+The helper handles HTTP requests to `[helper]/verify_start` as follows. It first
+checks for a well-formed `PAVerifyStartReq`. If not found, then the helper
+aborts and alerts the leader with "malformed verify start request".
+
+Next, it looks for PA parameters `PAParam` for which `PAVerifyStartReq.task ==
+PAParam.task`. If not found, then it aborts and alerts the leader with
+"unrecognized task".
+
+Next, it looks up the HPKE config and corresponding secret key associated with
+`PAVerifyStartReq.key_config_id`. If not found, then it aborts and alerts the
+leader with "unrecognized key config".
+
+Finally, for each pair of shares and message payloads, the helper does as
+follows. Let `share` denote the share. It computes the HPKE context as
+
+```
+context = SetupBaseR(share.enc, sk, PAParam.task)
+```
+
+where `sk` is the secret key corresponding to the HPKE key config. It then
+derives its proof share from `context` and `share.payload` as specified by the
+PA protocol. Finally, the helper uses the proof share and corresponding message
+to derive its response for the protocol run.
+
+The helper responds to the HTTP request from the leader with the following
+message:
+
+```
+struct {
+  PAProto proto;
+  bool resend_shares; // Whether the leader should resend shares
+  select (PAVerifyStartResp.proto) {
+    case prio: PrioVerifyStartResp payloads<0..2^16-1>;
+    case hits: HitsVerifyStartResp payloads<0..2^16-1>;
+  }
+} PAVerifyStartResp;
+```
+
+The i-th element of `PAVerifyStartResp.payloads` corresponds to the i-th element
+of `PAVerifyStartReq.payloads`. The boolean `resend_shares` is set to `true` if
+the helper wants the leader to retransmit its shares in the last step of the
+input-validation protocol. This allows the helper to implement the protocol
+statelessly, if desired.
+
+#### Verify Finish
+
+The leader handles the helper's response to its HTTP request by first checking
+for a well-formed `PAVerifyStartResp` message. If not found, then it aborts and
+alerts the helper with "malformed verify start response".
+
+**Request.**
+The leader sends an HTTP request to `[helper]/verify_finish` with the following
+message:
+
+```
+struct {
+  PATask task;                     // PAUpload.task
+  uint8 key_config_id;             // PAUpload.helper_key_config_id
+  PAHelperShare shares<0..2^24-1>; // Shares (may be empty)
```

Comment:
If this is empty then I think you still need a list of share unique identifiers so that helper knows how to resolve the list of `{Prio|Hits}VerifyFinishReq` messages it gets against the shares it has stored.

Reply:
That seems like a lot of overhead that's not really necessary. It should be clear that the first element of `payloads` corresponds to the first share, the second element of `payloads` to the second share, and so on. This should be fairly straightforward if the helper is already maintaining state.